impexp-core/src/main/java/org/citydb/core/operation/common/cache/CacheTableManager.java

```diff
@@ -27,241 +27,252 @@
-	private final Connection cacheConnection;
-	private final Config config;
-
-	private final ConcurrentHashMap<CacheTableModel, CacheTable> cacheTables;
-	private final ConcurrentHashMap<CacheTableModel, BranchCacheTable> branchCacheTables;
-
-	private String cacheDir;
-	private AbstractDatabaseAdapter databaseAdapter;
-	private Connection databaseConnection;
-
-	public CacheTableManager(int concurrencyLevel, Config config) throws SQLException, IOException {
-		if (config.getGlobalConfig().getCache().isUseDatabase()) {
-			cacheAdapter = DatabaseConnectionPool.getInstance().getActiveDatabaseAdapter();
-			cacheConnection = DatabaseConnectionPool.getInstance().getConnection();
-		}
-
-		else {
-			File tempDir = checkTempDir(config.getGlobalConfig().getCache().getLocalCachePath());
-			log.debug("Local cache directory is '" + tempDir.getAbsolutePath() + "'.");
-			cacheAdapter = new H2Adapter();
-
-			try {
-				Class.forName(cacheAdapter.getConnectionFactoryClassName());
-			} catch (ClassNotFoundException e) {
-				throw new SQLException(e);
-			}
-
-			try {
-				cacheDir = tempDir.getAbsolutePath() + File.separator + DefaultGMLIdManager.getInstance().generateUUID("");		
-				cacheConnection = DriverManager.getConnection(cacheAdapter.getJDBCUrl(cacheDir + File.separator + "tmp", -1, null), "sa", "");
-			} catch (SQLException e) {
-				deleteTempFiles(new File(cacheDir));
-				throw e;
-			}
-		}
+	private final String cacheDir;
+	private final Cache primaryCache;
 
-		cacheConnection.setAutoCommit(false);
+	private final Map<CacheMode, Cache> caches = new ConcurrentHashMap<>();
+	private final Map<CacheTableModel, CacheTable> cacheTables = new ConcurrentHashMap<>();
+	private final Map<CacheTableModel, BranchCacheTable> branchCacheTables = new ConcurrentHashMap<>();
 
-		cacheTables = new ConcurrentHashMap<>(CacheTableModel.values().length, 0.75f, concurrencyLevel);
-		branchCacheTables = new ConcurrentHashMap<>(CacheTableModel.values().length, 0.75f, concurrencyLevel);
-		this.config = config;
+	public CacheTableManager(org.citydb.config.project.global.Cache cacheConfig) throws SQLException, IOException {
+		cacheDir = cacheConfig.getLocalCachePath();
+		primaryCache = getOrCreateCache(cacheConfig.getCacheMode());
 	}
 
-	public AbstractDatabaseAdapter getDatabaseAdapter() {
-		return cacheAdapter;
+	public AbstractDatabaseAdapter getCacheAdapter() {
+		return primaryCache.adapter;
 	}
 
 	public CacheTable createCacheTable(CacheTableModel model) throws SQLException {
-		return createCacheTable(model, cacheConnection, cacheAdapter);		
+		return getOrCreateCacheTable(primaryCache, model, false, CacheTable.class);
 	}
 
-	public CacheTable createCacheTableInDatabase(CacheTableModel model) throws SQLException {
-		initDatabaseConnection();
-		return createCacheTable(model, databaseConnection, databaseAdapter);
+	public CacheTable createCacheTable(CacheTableModel model, CacheMode mode) throws SQLException {
```

Comment:
If we use the API like this:

```
CacheTable table1 = cacheTableManager.createCacheTable(CacheTableModel.ID_LIST, CacheMode.DATABASE);
CacheTable table2 = cacheTableManager.createCacheTable(CacheTableModel.ID_LIST, CacheMode.LOCAL);

```

In this case, the `table1` and `table2` are identical. It would be better that the second code line should throw an exception like "_cacheTableMode.ID_LIST shall not be created twice, even with different CacheMode_...."

Reply:
The behaviour is intended to be this way. Assume the `createCacheTable` method is invoked by two concurrent threads for the same CacheTableModel. Then both threads will receive a valid reference to the same table and write data to it. Otherwise, the creation of cache tables would always have to happen in a synchonized way.

Reply:
Good point. An exception is thrown now as suggested. 4b60ab9a3a3d0900c5dcbc4f4422eece601819a5